src/qengine/opencl.cpp

```diff
@@ -75,6 +78,22 @@ void QEngineOCL::Sync()
     UnlockSync();
 }
 
+void QEngineOCL::clFinish(bool doHard)
+{
+    if (device_context == NULL) {
+        return;
+    }
+
+    if (doHard) {
+        queue.finish();
+    } else {
+        for (unsigned int i = 0; i < (device_context->wait_events.size()); i++) {
+            device_context->wait_events[i].wait();
```

Comment:
Would [wait_group_events](https://www.khronos.org/registry/OpenCL/sdk/2.1/docs/man/xhtml/wait_group_events.html) be a better choice here?